ports/stm32/sdram.c

```diff
@@ -254,52 +253,113 @@ static void sdram_init_seq(SDRAM_HandleTypeDef
     #endif
 }
 
-bool sdram_test(bool fast) {
+void sdram_enter_low_power() {
+    // Enter self-refresh mode.
+    // In self-refresh mode the SDRAM retains data with external clocking.
+    FMC_SDRAM_CommandTypeDef command;
+    command.CommandMode = FMC_SDRAM_CMD_SELFREFRESH_MODE;
+    command.CommandTarget = FMC_SDRAM_CMD_TARGET_BANK;
+    command.AutoRefreshNumber = 1;
+    command.ModeRegisterDefinition = 0;
+    HAL_SDRAM_SendCommand(&hsdram, &command, 0xFFFF);
+}
+
+void sdram_leave_low_power() {
+    // Exit self-refresh mode.
+    // Self-refresh mode is exited when the device is accessed or the mode bits are
+    // set to Normal mode, so technically it's not necessary to call this functions.
+    FMC_SDRAM_CommandTypeDef command;
+    command.CommandMode = FMC_SDRAM_CMD_NORMAL_MODE;
+    command.CommandTarget = FMC_SDRAM_CMD_TARGET_BANK;
+    command.AutoRefreshNumber = 1;
+    command.ModeRegisterDefinition = 0;
+    HAL_SDRAM_SendCommand(&hsdram, &command, 0xFFFF);
+}
+
+bool __attribute__((optimize("O0"))) sdram_test(bool exhaustive) {
```

Comment:
@andrewleech what do you think about inverting the logic of this argument?

Reply:
Currently this function is run from main.c with the arg set to true: https://github.com/micropython/micropython/blob/master/ports/stm32/main.c#L399

On our hardware (F765, 32MB sdram) the `exhaustive` mode takes about a second to run - I suspect this might be even longer on slower boards like STM32F429DISC. 
I'm not sure it adds a lot of value either? The row/column style tests used otherwise have picked up manufacturing problems on a number of our early alpha boards run (pins not-connected / shorted).

Reply:
The logic has been inverted, with this change `sdram_test(true)` will run the "slow" version. This can be changed back while keeping the updated test, but I think "exhaustive" describes it better. 

> I'm not sure it adds a lot of value either? The row/column style tests used otherwise have picked up manufacturing problems on a number of our early alpha boards run (pins not-connected / shorted).

This updated version *should* be better at detecting refresh rate issues.

Reply:
Ah, sorry I missed that your additions to the "exhaustive" part of the test had that comment(was on mobile before), I definitely like the additions you've made to the tests overall! The inversion of the logic does describe it better, I'm just worried about startup speed increases.

Am I right in thinking refresh issues should be a hardware design time issue? As in, when proofing out a hardware design this will be useful in tweaking the settings, but once determined they should be stable? What do you think about enabling exhaustive on debug builds, but keeping it off for normal builds?

Reply:
Yes this is not meant to run in production, it can detect misconfigured refresh rates and/or power issues, because by the time it reads back the memory (with cache disabled) if there's an issue the contents will be lost/changed. Like I said I can invert back the logic or you can invert it in your main.c the next time you update micropython.

Reply:
Great, yeah I mean that inversion in main.c should be done in this MR as well. The usage of the function is automatically enabled in all builds with SDRAM configured, I don't have a separate main.c
Better still make it true/false in main.c depending on whether its a DEBUG build or not?

Reply:
Oh I see, yes I can fix that in main.c as well. As for debug builds, it will slow down the firmware when most of the time you're likely debugging something else, so I think it's best to just leave it as false by default, and change it explicitly when needed.

Reply:
Yeah true. It could also be made a `MICROPY_HW_SDRAM_EXTENDED_TEST` type config. Might not be huge value in that, other than a slightly more visible marker to other developers that it's something that might be of interest.

Reply:
@andrewleech @dpgeorge I inverted the test in main.c and forced a push, then in a separate commit I put to control the test in board config files.